doc/how-to-freeze-versions.rst

```diff
@@ -0,0 +1,247 @@
+    contains, don't include a ``.freeze`` file, don't add it to any field of the
+    package description that would have ``cabal sdist`` include it in the
+    ``.tar.gz``. In general, don't include anything in the package description
+    that relates to the project environment, like ``cabal.project`` or
+    ``cabal.project.local``.
+
+Freezing workflows
+^^^^^^^^^^^^^^^^^^
+
+.. Warning::
+    For each of these workflows, you may have to first delete the
+    ``index-state`` line from ``cabal.project`` (and from
+    ``cabal.project.freeze`` if it exists) and then run ``cabal update`` to
+    ensure that cabal will have newer versions to re-resolve the dependencies
+    with. Alternatively, you can run ``cabal update
+    --ignore-project``.
+
+Freeze
+    If the ``cabal.project.freeze`` file doesn't exist, generating one is a
+    great way to see what versions of dependencies are currently being used even
+    if you choose to discard the ``.freeze`` file after inspecting it.
+
+Thaw, Freeze
+    If you changed the version ranges of any of the dependencies in any of your
+    project's package descriptions, in any ``.cabal`` file, then delete the
+    ``cabal.project.freeze`` file if it already exists and run ``cabal freeze``
+    to generate fresh version of ``cabal.project.freeze``.
+
+Freeze, Freeze (Freezing Harder)
+    You might in some cases want to skip deletion of ``cabal.project.freeze``,
+    but keep in mind that in that case ``cabal freeze`` will use existing
+    ``cabal.project.freeze`` when resolving dependencies, therefore not updating
+    any existing dependencies, only adding new ones.
+
+Partial Thaw, Freeze
+    If you want to a pick up a different version of a single dependency, you can
+    delete its constraint from ``cabal.project.freeze`` and then run ``cabal
+    freeze`` again.
+
+.. Note::
+
+    If not sure, pick the "thaw, freeze" workflow, as it is the safest, the
+    simplest and the most common. Finally, you will always want to commit the
+    changed ``cabal.project.freeze`` to version control.
+
+Ensuring everything is frozen
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. Note::
+
+    If the ``.freeze`` file already has version equality constraints for every
+    package that is a dependency of the project, then the solver will not be
+    able to find a different version for any of them, the ``.freeze`` file
+    cannot change and, at that point when every dependency is frozen, ``cabal
+    freeze`` becomes an idempotent operation.
+
+Adding a dependency to one of the packages in a project without freezing harder
+leaves the newly added dependency susceptible to getting updated unexpectedly
+when the solver can find a different version for it.  Running ``cabal freeze``
+will show this vulnerability to a human or an automated check that notices a new
+version equality constraint in the ``.freeze`` file, a constraint for a package
+that wasn't in the ``.freeze`` file before.
```

Comment:
I would suggest rephrasing this paragraph along these lines

> When adding a new dependency to a project that utilizes a freeze file, it is often desirable to regenerate the freeze file using one of the methods described above. This ensures that the freeze file contains a constraint for the new dependency.

If we want to mention the idea of an automated check, perhaps explicitly suggest how one might work by providing shell commands that implement it.